contracts/src/CrossChainApplications/NativeTokenBridge/ERC20TokenSource.sol

```diff
@@ -0,0 +1,174 @@
+
+        require(
+            erc20ContractAddress_ != address(0),
+            "ERC20TokenSource: Invalid ERC20 Contract Address"
+        );
+        erc20ContractAddress = erc20ContractAddress_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "ERC20TokenSource: Unauthorized TeleporterMessenger contract"
+        );
+
+        // Only allow messages from the destination chain.
+        require(
+            senderBlockchainID == destinationBlockchainID,
+            "ERC20TokenSource: Invalid Destination Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(
+            senderAddress == nativeTokenDestinationAddress,
+            "ERC20TokenSource: Unauthorized Sender"
+        );
+
+        (address recipient, uint256 amount) = abi.decode(
+            message,
+            (address, uint256)
+        );
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero Recipient Address"
+        );
+
+        // Transfer to recipient
+        IERC20(erc20ContractAddress).safeTransfer(recipient, amount);
+
+        emit UnlockTokens(recipient, amount);
+    }
+
+    /**
+     * @dev See {IERC20TokenSource-transferToDestination}.
+     */
+    function transferToDestination(
+        address recipient,
+        uint256 totalAmount,
+        uint256 feeAmount,
+        address[] calldata allowedRelayerAddresses
+    ) external nonReentrant {
+        // The recipient cannot be the zero address.
+        require(
+            recipient != address(0),
+            "ERC20TokenSource: zero Recipient Address"
+        );
+
+        // Lock tokens in this contract. Supports "fee/burn on transfer" ERC20 token
+        // implementations by only bridging the actual balance increase reflected by the call
+        // to transferFrom.
+        uint256 adjustedAmount = SafeERC20TransferFrom.safeTransferFrom(
+            IERC20(erc20ContractAddress),
+            totalAmount
+        );
+
+        // Ensure that the adjusted amount is greater than the fee to be paid.
+        require(
+            adjustedAmount > feeAmount,
+            "ERC20TokenSource: insufficient adjusted amount"
+        );
+
+        // Allow the Teleporter messenger to spend the fee amount.
+        if (feeAmount > 0) {
+            IERC20(erc20ContractAddress).safeIncreaseAllowance(
+                address(teleporterMessenger),
+                feeAmount
+            );
+        }
+
+        uint256 transferAmount = totalAmount - feeAmount;
```

Comment:
I think the `transferAmount` here should be `adjustedAmount - feeAmount`, right?